File: src/clj/pdp/core/conf.clj
;; Copyright (c) Konrad Grzanek. All rights reserved.
;; Created 2015-11-19

(in-ns 'pdp.core)

(defn make-conf [n] (jpdp.core.Conf. (int n)))

(defmacro conf-set!
  [conf prop value]
  (let [conf (vary-meta conf assoc :tag 'jpdp.core.Conf)]
    `(.set ~conf (lv ~prop) (lv ~value))))


(defmacro conf-get
  [conf prop]
  (let [conf (vary-meta conf assoc :tag 'jpdp.core.Conf)]
    `(.get ~conf (lv ~prop))))


;; (defprops Engine Gearbox)
;; (defvals  Diesel Benz Auto Manual)
;; (def c (make-conf 100))

File: src/java/jpdp/core/Conf.java
/*
 * Copyright (c) Konrad Grzanek. All rights reserved.
 * Created 2015-11-19
 */
package jpdp.core;

public final class Conf {

  private final Value[] values;

  public Conf(int n) {
    values = new Value[n];
  }

  public synchronized Value get(Prop prop) {
    return values[prop.code];
  }

  public synchronized void set(Prop prop, Value value) {
    values[prop.code] = value;
  }
}
